File: objectModel/TypeScript/Persistence/CdmFolder/EntityAttributePersistence.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { CdmFolder } from '..';
import {
    CdmCorpusContext,
    CdmEntityAttributeDefinition,
    CdmEntityReference,
    cdmLogCode,
    cdmObjectType,
    CdmTraitGroupReference,
    CdmTraitReference,
    CdmTraitReferenceBase,
    copyOptions,
    Logger,
    resolveOptions
} from '../../internal';
import * as copyDataUtils from '../../Utilities/CopyDataUtils';
import { ProjectionPersistence } from './Projections/ProjectionPersistence';
import { AttributeResolutionGuidance, EntityAttribute, EntityReferenceDefinition, PurposeReference, TraitGroupReference, TraitReference } from './types';
import * as utils from './utils';

export class EntityAttributePersistence {
    private static TAG: string = EntityAttributePersistence.name;
    
    public static fromData(ctx: CdmCorpusContext, object: EntityAttribute): CdmEntityAttributeDefinition {
        const entityAttribute: CdmEntityAttributeDefinition = ctx.corpus.MakeObject(cdmObjectType.entityAttributeDef, object.name);

        entityAttribute.description = utils.propertyFromDataToString(object.description);
        entityAttribute.displayName = utils.propertyFromDataToString(object.displayName);
        entityAttribute.explanation = utils.propertyFromDataToString(object.explanation);

        entityAttribute.cardinality = utils.cardinalitySettingsFromData(object.cardinality, entityAttribute);

        entityAttribute.isPolymorphicSource = object.isPolymorphicSource;

        if (object.entity && typeof(object.entity) !== 'string' && 'source' in object.entity) {
            const inlineEntityRef: CdmEntityReference = ctx.corpus.MakeObject<CdmEntityReference>(cdmObjectType.entityRef, undefined);
            inlineEntityRef.explicitReference = ProjectionPersistence.fromData(ctx, object.entity);
            entityAttribute.entity = inlineEntityRef;
        } else {
            entityAttribute.entity = CdmFolder.EntityReferencePersistence.fromData(ctx, object.entity);
        }

        entityAttribute.purpose = CdmFolder.PurposeReferencePersistence.fromData(ctx, object.purpose);
        utils.addArrayToCdmCollection<CdmTraitReferenceBase>(
            entityAttribute.appliedTraits,
            utils.createTraitReferenceArray(ctx, object.appliedTraits)
        );

        // Ignore resolution guidance if the entity is a projection
        if (object.resolutionGuidance && object.entity && typeof(object.entity) !== 'string' && 'source' in object.entity) {
            Logger.error(ctx, this.TAG, this.fromData.name, null, cdmLogCode.ErrPersistEntityAttrUnsupported, entityAttribute.name);
        } else {
            entityAttribute.resolutionGuidance =
                CdmFolder.AttributeResolutionGuidancePersistence.fromData(ctx, object.resolutionGuidance);
        }

        return entityAttribute;
    }

    public static toData(instance: CdmEntityAttributeDefinition, resOpt: resolveOptions, options: copyOptions): EntityAttribute {
        let entity: (string | EntityReferenceDefinition);
        entity = instance.entity ? instance.entity.copyData(resOpt, options) as (string | EntityReferenceDefinition) : undefined;
        const appliedTraits: CdmTraitReferenceBase[] = instance.appliedTraits ?
            instance.appliedTraits.allItems.filter(
                (trait: CdmTraitReferenceBase) => trait instanceof CdmTraitGroupReference || !(trait as CdmTraitReference).isFromProperty) : undefined;

        const object: EntityAttribute = {
            name: instance.name,
            description: instance.description,
            displayName: instance.displayName,
            explanation: instance.explanation,
            isPolymorphicSource: instance.isPolymorphicSource,
            purpose: instance.purpose
            ? instance.purpose.copyData(resOpt, options) as (string | PurposeReference)
                : undefined,
            entity: entity,
            appliedTraits: copyDataUtils.arrayCopyData<string | TraitReference | TraitGroupReference>(resOpt, appliedTraits, options),
            resolutionGuidance: instance.resolutionGuidance
                ? instance.resolutionGuidance.copyData(resOpt, options) as AttributeResolutionGuidance : undefined,
            cardinality: utils.cardinalitySettingsToData(instance.cardinality)
        };

        return object;
    }
}

File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Persistence/Syms/EntityAttributePersistence.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Persistence.Syms
{
    using Microsoft.CommonDataModel.ObjectModel.Cdm;
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Persistence.Syms.Types;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using Microsoft.CommonDataModel.ObjectModel.Utilities.Logging;
    using Newtonsoft.Json.Linq;
    using System.Linq;

    class EntityAttributePersistence
    {
        private static readonly string Tag = nameof(EntityAttributePersistence);

        public static CdmEntityAttributeDefinition FromData(CdmCorpusContext ctx, JToken obj)
        {
            var entityAttribute = ctx.Corpus.MakeObject<CdmEntityAttributeDefinition>(CdmObjectType.EntityAttributeDef, (string)obj["name"]);

            entityAttribute.Description = Utils.PropertyFromDataToString(obj["description"]);
            entityAttribute.DisplayName = Utils.PropertyFromDataToString(obj["displayName"]);
            entityAttribute.Explanation = Utils.PropertyFromDataToString(obj["explanation"]);

            if (obj["cardinality"] != null)
            {
                string minCardinality = null;
                if (obj["cardinality"]["minimum"] != null)
                    minCardinality = (string)obj["cardinality"]["minimum"];

                string maxCardinality = null;
                if (obj["cardinality"]["maximum"] != null)
                    maxCardinality = (string)obj["cardinality"]["maximum"];

                if (string.IsNullOrWhiteSpace(minCardinality) || string.IsNullOrWhiteSpace(maxCardinality))
                    Logger.Error((ResolveContext)ctx, Tag, nameof(FromData), null, CdmLogCode.ErrPersistCardinalityPropMissing);

                if (!CardinalitySettings.IsMinimumValid(minCardinality))
                    Logger.Error((ResolveContext)ctx, Tag, nameof(FromData), null, CdmLogCode.ErrPersistInvalidMinCardinality, minCardinality);

                if (!CardinalitySettings.IsMaximumValid(maxCardinality))
                    Logger.Error((ResolveContext)ctx, Tag, nameof(FromData), null, CdmLogCode.ErrPersistInvalidMaxCardinality, maxCardinality);

                if (!string.IsNullOrWhiteSpace(minCardinality) &&
                    !string.IsNullOrWhiteSpace(maxCardinality) &&
                    CardinalitySettings.IsMinimumValid(minCardinality) &&
                    CardinalitySettings.IsMinimumValid(maxCardinality))
                {
                    entityAttribute.Cardinality = new CardinalitySettings(entityAttribute)
                    {
                        Minimum = minCardinality,
                        Maximum = maxCardinality
                    };
                }
            }

            entityAttribute.IsPolymorphicSource = (bool?)obj["isPolymorphicSource"];

            bool isProjection = obj["entity"] != null &&
                !(obj["entity"] is JValue) &&
                obj["entity"]["source"] != null;
            if (isProjection)
            {
                Logger.Warning((ResolveContext)ctx, Tag, nameof(FromData), null, CdmLogCode.WarnPersistSymsProjNotExist);
            }

            entityAttribute.Entity = EntityReferencePersistence.FromData(ctx, obj["entity"]);

            entityAttribute.Purpose = PurposeReferencePersistence.FromData(ctx, obj["purpose"]);
            Utils.AddListToCdmCollection(entityAttribute.AppliedTraits, Utils.CreateTraitReferenceList(ctx, obj["appliedTraits"]));
            // ignore resolution guidance if the entity is a projection
            if (obj["resolutionGuidance"] != null && isProjection)
            {
                Logger.Error((ResolveContext)ctx, Tag, nameof(FromData), null, CdmLogCode.ErrPersistEntityAttrUnsupported, entityAttribute.Name);
            }
            else
            {
                entityAttribute.ResolutionGuidance = AttributeResolutionGuidancePersistence.FromData(ctx, obj["resolutionGuidance"]);
            }
            return entityAttribute;
        }

        public static EntityAttribute ToData(CdmEntityAttributeDefinition instance, ResolveOptions resOpt, CopyOptions options)
        {
            return new EntityAttribute
            {
                Explanation = instance.Explanation,
                Name = instance.Name,
                IsPolymorphicSource = instance.IsPolymorphicSource,
                Entity = Utils.JsonForm(instance.Entity, resOpt, options),
                Purpose = Utils.JsonForm(instance.Purpose, resOpt, options),
                AppliedTraits = CopyDataUtils.ListCopyData(resOpt, instance.AppliedTraits?
                    .Where(trait => trait is CdmTraitGroupReference || !(trait as CdmTraitReference).IsFromProperty), options),
                ResolutionGuidance = Utils.JsonForm(instance.ResolutionGuidance, resOpt, options),
                DisplayName = instance.GetProperty("displayName"),
                Description = instance.GetProperty("description")
            };
        }
    }
}
